refresh.template.py

```diff
@@ -979,20 +979,52 @@ def _get_commands(target: str, flags: str):
 
 
     # Then, actually query Bazel's compile actions for that configured target
-    target_statement = f'deps({target})' # TODO we should always be quoting targets when we splice them in. Let's use single quotes like with mnemonic, above. See slightly down in https://bazel.build/query/language#tokens
+    target_statement = f"deps('{target}')" # TODO we should always be quoting targets when we splice them in. Let's use single quotes like with mnemonic, above. See slightly down in https://bazel.build/query/language#tokens
     compile_commands = [] # TODO simplify loop, especially if we can reduce it to one command per case (see below)? Move warning messages outside?
     if file_flags:
         file_path = file_flags[0]
+        # Normalize file_path into 2 types of relative paths.
+        # Source file. 1. <WORKSPACE_ROOT>/srcs/file 2. <OUTPUT_BASE>/external/<REPO>/srcs/file
+        # Genfiles or output file. <EXECUTION_ROOT>/bazel-out/<configurations>/bin/srcs/file
+
+        pure_path = pathlib.PurePath(file_path)
+
+        if pure_path.is_absolute():
+            workspace_root = subprocess.check_output(["bazel", "info", "workspace"], stderr=subprocess. DEVNULL).decode("utf-8").strip()
+            output_base = subprocess.check_output(["bazel", "info", "output_base"], stderr=subprocess. DEVNULL).decode("utf-8").strip()
+            execution_root = subprocess.check_output(["bazel", "info", "execution_root"], stderr=subprocess. DEVNULL).decode("utf-8").strip()
+
+            if pure_path.is_relative_to(workspace_root):
+                file_path = str(pure_path.relative_to(workspace_root))
+            # Execution_root is longer then output_base
+            elif pure_path.is_relative_to(execution_root):
+                file_path = str(pure_path.relative_to(execution_root))
+            elif pure_path.is_relative_to(output_base):
+                file_path = str(pure_path.relative_to(output_base))
+            else:
+                # Treat pure_path as system absolute path
+                pass
```

Comment:
Really appreciate your tackling. But on a quick read, I think this part could use some polish along the lines we'd discussed.

> A couple things that jump out:
> 
> We should probably still have an error for the case that's in-output-base but not external or execroot.
> ...
> (Remember the backwards compatible relative_to helper, as above)
> (I'm sure there's more I missed on a quick pass; I'm counting on you to experiment and track them down, and give it some solid manual testing.)

Probably also worth using subprocess.run instead of subprocess.check_output, just because of deprecation and to match the others? (Also, there's the encoding issue.)

Reply:
I would like to ask if there are any other problems with the processing of paths besides subprocess.check_output? At present, I have modified these 3 calls, but I have not changed other places in the file